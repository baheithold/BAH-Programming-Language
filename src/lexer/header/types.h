/*
 *  Author:         Brett Heithold
 *  File:           types.h
 *  Created:        10/21/2020
 *  Last Revised:   11/1/2020
 */

#ifndef __TYPES_H__
#define __TYPES_H__


/* Types */
extern char *INTEGER_TYPE;
extern char *REAL_TYPE;
extern char *STRING_TYPE;
extern char *BOOLEAN_TYPE;
extern char *ID_TYPE;
extern char *VARIABLE_TYPE;
extern char *NULL_TYPE;

/* Punctuation */
extern char *OPAREN;
extern char *CPAREN;
extern char *OBRACE;
extern char *CBRACE;
extern char *OBRACKET;
extern char *CBRACKET;
extern char *COMMA;
extern char *COLON;
extern char *SEMICOLON;

/* Binary Operators */
extern char *PLUS_BINARY;
extern char *MINUS_BINARY;
extern char *TIMES_BINARY;
extern char *DIVIDE_BINARY;
extern char *POW_BINARY;
extern char *MODULO_BINARY;
extern char *ASSIGN_BINARY;
extern char *PLUS_ASSIGN_BINARY;
extern char *MINUS_ASSIGN_BINARY;
extern char *TIMES_ASSIGN_BINARY;
extern char *DIVIDE_ASSIGN_BINARY;
extern char *POW_ASSIGN_BINARY;
extern char *MODULO_ASSIGN_BINARY;
extern char *DOT_BINARY;

/* Unary Operators */
extern char *INCREMENT_UNARY;
extern char *DECREMENT_UNARY;
extern char *NEGATE_UNARY;

/* Comparators */
extern char *EQUALS_COMPARATOR;
extern char *NOT_EQUALS_COMPARATOR;
extern char *GREATER_THAN_COMPARATOR;
extern char *LESSER_THAN_COMPARATOR;
extern char *GREATER_EQUALS_COMPARATOR;
extern char *LESSER_EQUALS_COMPARATOR;

/* Logical Operators */
extern char *AND;
extern char *OR;
extern char *XOR;

/* Keywords */
extern char *IF;
extern char *ELSE;
extern char *FOR;
extern char *WHILE;
extern char *FUNCTION;
extern char *VAR;
extern char *RETURN;
extern char *BREAK;
extern char *CONTINUE;
extern char *CLASS;
extern char *LAMBDA;
extern char *VOID;
extern char *IMPORT;
extern char *PRINT;
extern char *PRINTLN;

/* Other */
extern char *UNKNOWN;
extern char *BAD_NUMBER;
extern char *BAD_CHARACTER;
extern char *BAD_STRING;
extern char *END_OF_INPUT;
extern char *VARIABLE_DEFINITION;
extern char *FUNCTION_DEFINITION;
extern char *FUNCTION_CALL;
extern char *CLASS_DEFINITION;
extern char *ENVIRONMENT;
extern char *JOIN;


#endif // !__TYPES_H__
