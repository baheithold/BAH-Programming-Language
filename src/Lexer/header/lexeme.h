/*
 *  Author:         Brett Heithold
 *  File:           lexeme.h
 *  Created:        10/21/2020
 *  Last Revised:   10/24/2020
 */

#ifndef __LEXEME_H__
#define __LEXEME_H__


#include <stdio.h>


typedef struct Lexeme Lexeme;

extern Lexeme   *newLexeme(char *, char *);
extern char     *getLexemeType(Lexeme *);
extern void      setLexemeType(Lexeme *, char *);
extern int       getLexemeIntegerValue(Lexeme *);
extern double    getLexemeRealValue(Lexeme *);
extern char     *getLexemeStringValue(Lexeme *);
extern int       getLexemeLineNumber(Lexeme *);
extern void      setLexemeLineNumber(Lexeme *, int);
extern void      printLexeme(FILE *, Lexeme *);

#endif // !__LEXEME_H__
