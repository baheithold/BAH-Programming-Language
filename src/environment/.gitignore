# Object files (directory)
obj

# Executables
test-env

# Directories
actual
