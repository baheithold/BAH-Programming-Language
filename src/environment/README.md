# Environments

## Implementing environments and scope
The purpose of an environment is two fold: to hold the bindings between variables and their values and to implement scope. The environment is implemented internally using a list of parallel lists. A new environment is created using the `createEnvironment` function. The new environment created by `createEnvironment` has no bindings and is represented internally as two empty lists in a list.

** The use of `table` in this document refers to the list of parallel lists used in the environment structure to store the bindings between identifiers and values;

The environment subsystem provides six basic operation:
* `createEnvironment` creates and returns an empty environment. The new empty environment consists of one empty table in the list of tables.
* `lookupEnvironment` searches for a variable starting from the most local table and proceeding to the most outer table. The value of the first occurrence of the queried variable is returned.
* `updateEnvironment` rebinds a new value to the queried identifier. The value of the first occurrence of the queried identifier is updated to the new value. The new value is returned by the function.
* `insertEnvironment` inserts an identifier and a value into the most local table. This function returns the value inserted.
* `extendEnvironment` extends a given environment with a list of identifiers and values. The new environment created by the function is returned with the identifiers and values placed in a table at the head of the given environment. The given environment remains unmodified.
* `printEnvironment` prints the environment to a file pointer. A flag parameter is provided that allows the caller to set whether the function only prints the local environment or prints the entired enclosing environment.

## Looking up and updating variables
Looking up the value of an identifier is as simple as finding the identifier/value binds within the tables. The tables are searched for the corresponding identifier from the most local table to the most outer table. If no identifier is found, then an undefined identifier error is thrown.

Updating the value of an identifier works in the same way as looking up a value except that the value is changed in the update function.

## Inserting a new variable
A new identifier and value binding is inserted into the most local environment at the time of definition. Internally, the most local environment is the first table (the first two lists) in the environments list. The new binding is added to the head of the parallel lists. An advantage of adding new bindings to the head of the parallel lists, is that this placement allows for redefinitions of identifiers to shadow previous definitions within the same scope.

## Extending an environment
Extending an environment takes place when nesting scopes inside of other scopes. This might happen when, for instance, functions are defined with formal parameters. Environment extension is accomplished by first creating a new empty environment. The next step is to populate the new environment with the identifier/value bindings defined in the new environment. The final step is to link the new environment with its enclosing environment.
