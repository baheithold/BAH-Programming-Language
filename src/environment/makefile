CC 			 = gcc
SRCDIR 		 = src
HEADERDIR 	 = header
OBJDIR 		 = obj
LEXERPATH 	 = ../lexer
TESTDIR 	 = testing
EXPECTEDPATH = $(TESTDIR)/expected
ACTUALPATH 	 = $(TESTDIR)/actual
OBJS  	  	 = $(OBJDIR)/test-env.o $(OBJDIR)/environment.o $(OBJDIR)/types.o $(OBJDIR)/lexeme.o
OOPTS 		 = -Wall -Wextra -g -c
LOPTS 		 = -Wall -Wextra -g


all : test-env


test-env : $(OBJS)
	@echo Creating test-env executable...
	@$(CC) $(LOPTS) $(OBJDIR)/* -o test-env
	@echo Done.


$(OBJDIR)/test-env.o : $(TESTDIR)/test-env.c $(HEADERDIR)/environment.h $(LEXERPATH)/$(HEADERDIR)/types.h $(LEXERPATH)/$(HEADERDIR)/lexeme.h
	@echo Compiling test-env...
	@mkdir -p $(OBJDIR)
	@$(CC) $(OOPTS) $(TESTDIR)/test-env.c -o $@


$(OBJDIR)/types.o : $(LEXERPATH)/$(SRCDIR)/types.c $(LEXERPATH)/$(HEADERDIR)/types.h
	@echo Compiling types...
	@mkdir -p $(OBJDIR)
	@$(CC) $(OOPTS) $(LEXERPATH)/$(SRCDIR)/types.c -o $@


$(OBJDIR)/lexeme.o : $(LEXERPATH)/$(SRCDIR)/lexeme.c $(LEXERPATH)/$(HEADERDIR)/lexeme.h $(LEXERPATH)/$(HEADERDIR)/types.h
	@echo Compiling lexeme...
	@mkdir -p $(OBJDIR)
	@$(CC) $(OOPTS) $(LEXERPATH)/$(SRCDIR)/lexeme.c -o $@

$(OBJDIR)/environment.o : $(SRCDIR)/environment.c $(HEADERDIR)/environment.h $(LEXERPATH)/$(HEADERDIR)/lexeme.h $(LEXERPATH)/$(HEADERDIR)/types.h
	@echo Compiling environment...
	@mkdir -p $(OBJDIR)
	@$(CC) $(OOPTS) $(SRCDIR)/environment.c -o $@

test : test-env
	@echo Running $@...
	@mkdir -p $(ACTUALPATH)
	@./test-env


rebuild : all
	@echo Rebuilding test-env...
	@make clean -s
	@make -s


clean :
	@echo Cleaning...
	@rm -rf $(OBJDIR) $(ACTUALPATH) test-env
