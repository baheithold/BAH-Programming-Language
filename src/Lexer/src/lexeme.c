/*
 *  Author:         Brett Heithold
 *  File:           lexeme.c
 *  Created:        10/21/2020
 *  Last Revised:   10/23/2020
 */


#include "../header/lexeme.h"
#include "../header/types.h"


#include <assert.h>
#include <stdlib.h>


struct Lexeme {
    char    *type;
    char    *stringValue;
    int      integerValue;
    double   realValue;
    int      lineNumber;
};


/********** Public Function Definitions **********/

Lexeme *newLexeme(char *type, char *value) {
    assert(type != NULL);
    Lexeme *result = malloc(sizeof(Lexeme));
    assert(result != NULL);
    result->type = type;
    if (type == INTEGER_TYPE) {
        result->integerValue = atoi(value);
    }
    else if (type == REAL_TYPE) {
        result->realValue = atof(value);
    }
    else {
        result->stringValue = value;
    }
    result->lineNumber = 0;
    return result;
}


char *getLexemeType(Lexeme *lexeme) {
    assert(lexeme != NULL);
    return lexeme->type;
}


int getLexemeIntegerValue(Lexeme *lexeme) {
    assert(lexeme != NULL);
    return lexeme->integerValue;
}


double getLexemeRealValue(Lexeme *lexeme) {
    assert(lexeme != NULL);
    return lexeme->realValue;
}


char *getLexemeStringValue(Lexeme *lexeme) {
    assert(lexeme != NULL);
    return lexeme->stringValue;
}


int getLexemeLineNumber(Lexeme *lexeme) {
    assert(lexeme != NULL);
    return lexeme->lineNumber;
}


void setLexemeLineNumber(Lexeme *lexeme, int line) {
    assert(lexeme != NULL);
    lexeme->lineNumber = line;
}


void printLexeme(FILE *fp, Lexeme *lexeme) {
    assert(lexeme != NULL);

    if (lexeme->type == INTEGER_TYPE) {
        // Print INTEGER_TYPE Lexeme
        fprintf(fp, "%s %d", lexeme->type, lexeme->integerValue);
    }
    else if (lexeme->type == REAL_TYPE) {
        // Print REAL_TYPE Lexeme
        fprintf(fp, "%s %f", lexeme->type, lexeme->realValue);
    }
    else if (lexeme->type == STRING_TYPE) {
        // Print REAL_TYPE Lexeme
        fprintf(fp, "%s \"%s\"", lexeme->type, lexeme->stringValue);
    }
    else {
        fprintf(fp, "%s", lexeme->type);
        if (lexeme->stringValue != NULL) {
            fprintf(fp, " %s", lexeme->stringValue);
        }
    }
}
