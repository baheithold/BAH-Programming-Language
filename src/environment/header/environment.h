/*
 *  Author:         Brett Heithold
 *  File:           environment.h
 *  Created on:     10/31/2020
 *  Last revision:  11/1/2020
 */

#ifndef __ENVIRONMENT_H__
#define __ENVIRONMENT_H__


#include "../../lexer/header/lexeme.h"

#include <stdbool.h>
#include <stdio.h>


extern Lexeme *createEnvironment(void);
extern Lexeme *lookupEnvironment(Lexeme *, Lexeme *);
extern Lexeme *updateEnvironment(Lexeme *, Lexeme *, Lexeme *);
extern Lexeme *insertEnvironment(Lexeme *, Lexeme *, Lexeme *);
extern Lexeme *extendEnvironment(Lexeme *, Lexeme *, Lexeme *);
extern void    printEnvironment(FILE *, Lexeme *, bool);


#endif // !__ENVIRONMENT_H__
