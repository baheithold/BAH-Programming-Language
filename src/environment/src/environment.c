/*
 *  Author:         Brett Heithold
 *  File:           environment.c
 *  created on:     10/31/2020
 *  Last Revision:  11/7/2020
 */


#include "../header/environment.h"
#include "../../lexer/header/lexeme.h"
#include "../../lexer/header/types.h"

#include <assert.h>
#include <stdbool.h>
#include <string.h>


#define DEBUG false


/********** Private Function Prototypes **********/
static bool    isSameVariable(Lexeme *, Lexeme *);


/********** Public Function Definitions **********/

Lexeme *createEnvironment(void) {
    if (DEBUG) {
        fprintf(stdout, "CALL: createEnvironment\n");
    }
    return extendEnvironment(NULL, NULL, NULL);
}

Lexeme *lookupEnvironment(Lexeme *variable, Lexeme *environment) {
    assert(variable != NULL);
    assert(environment != NULL);
    if (DEBUG) {
        fprintf(stdout, "CALL: lookupEnvironment\n");
    }

    while (environment != NULL) {
        Lexeme *vars = car(environment);
        Lexeme *vals = cadr(environment);
        while (vars != NULL) {
            if (isSameVariable(variable, car(vars))) {
                return car(vals);
            }
            vars = cdr(vars);
            vals = cdr(vals);
        }
        environment = cdr(cdr(environment));
    }
    fprintf(stdout, "ERROR: %s is undefined\n", getLexemeStringValue(variable));
    return NULL;
}

Lexeme *updateEnvironment(Lexeme *variable, Lexeme *value, Lexeme *environment) {
    assert(variable != NULL);
    assert(value != NULL);
    assert(environment != NULL);

    if (DEBUG) {
        fprintf(stdout, "CALL: updateEnvironment\n");
    }

    while (environment != NULL) {
        Lexeme *vars = car(environment);
        Lexeme *vals = cadr(environment);
        while (vars != NULL) {
            if (isSameVariable(variable, car(vars))) {
                return set_car(vals, value);
            }
            vars = cdr(vars);
            vals = cdr(vals);
        }
        environment = cdr(cdr(environment));
    }
    fprintf(stdout, "ERROR: %s is undefined\n", getLexemeStringValue(variable));
    return NULL;
}

Lexeme *insertEnvironment(Lexeme *variable, Lexeme *value, Lexeme *environment) {
    assert(variable != NULL);
    assert(value != NULL);
    assert(environment != NULL);

    if (DEBUG) {
        fprintf(stdout, "CALL: insertEnvironment\n");
    }

    set_car(environment, cons(JOIN, variable, car(environment)));
    set_car(cdr(environment), cons(JOIN, value, cadr(environment)));
    return value;
}

Lexeme *extendEnvironment(Lexeme *variables, Lexeme *values, Lexeme *environment) {
    if (DEBUG) {
        fprintf(stdout, "CALL: extendEnvironment\n");
    }

    return cons(ENVIRONMENT, variables, cons(ENVIRONMENT, values, environment));
}

void printEnvironment(FILE *fp, Lexeme *environment, bool local) {
    assert(environment != NULL);

    if (DEBUG) {
        fprintf(stdout, "CALL: printEnvironment\n");
    }

    int scopeLevel = 0;

    if (local) {
        fprintf(fp, "The local environment is:\n");
    }
    else {
        fprintf(fp, "The environment is:\n");
    }

    while (environment != NULL) {
        fprintf(fp, "Scope Level: %d\n", scopeLevel);
        Lexeme *vars = car(environment);
        Lexeme *vals = cadr(environment);
        while (vars != NULL) {
            fprintf(fp, "%s : ", getLexemeStringValue(car(vars)));
            printLexeme(fp, car(vals));
            fprintf(fp, "\n");
            vars = cdr(vars);
            vals = cdr(vals);
        }
        if (local) break;
        scopeLevel++;
        environment = cdr(cdr(environment));
    }
    fprintf(fp, "\n");
}


/********** Private Function Definitions **********/

bool isSameVariable(Lexeme *var1, Lexeme *var2) {
    assert(var1 != NULL);
    assert(var2 != NULL);

    if (DEBUG) {
        fprintf(stdout, "CALL: isSameVariable\n");
    }

    if (strcmp(getLexemeStringValue(var1), getLexemeStringValue(var2)) == 0) {
        return true;
    }
    return false;
}
