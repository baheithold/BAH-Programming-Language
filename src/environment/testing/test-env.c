/*
 *  Author:         Brett Heithold
 *  File:           test-env.c
 *  Created on:     10/31/2020
 *  Last revision:  11/2/2020
 */


#include "../header/environment.h"
#include "../../lexer/header/lexeme.h"
#include "../../lexer/header/types.h"


int main(void) {
    // Create environment
    Lexeme *global = createEnvironment();
    printf("Creating new environment...\n");


    // Inserting variables into the global scope
    Lexeme *x = newLexeme(ID_TYPE, "x");
    Lexeme *one = newLexeme(INTEGER_TYPE, "1");
    Lexeme *y = newLexeme(ID_TYPE, "y");
    Lexeme *two = newLexeme(INTEGER_TYPE, "2");
    Lexeme *z = newLexeme(ID_TYPE, "z");
    Lexeme *three = newLexeme(INTEGER_TYPE, "3");
    printf("Adding variable x with value 1\n");
    insertEnvironment(x, one, global);
    printf("Adding variable y with value 2\n");
    insertEnvironment(y, two, global);
    printf("Adding variable z with value 3\n");
    insertEnvironment(z, three, global);
    printEnvironment(stdout, global, false);

    // Extend environment
    printf("Extending the environment with w:4 and u:\"hello\"\n");
    Lexeme *w = newLexeme(ID_TYPE, "w");
    Lexeme *four = newLexeme(INTEGER_TYPE, "4");
    Lexeme *u = newLexeme(ID_TYPE, "u");
    Lexeme *hello = newLexeme(STRING_TYPE, "hello");
    Lexeme *envLocal_01 = extendEnvironment(NULL, NULL, global);
    insertEnvironment(w, four, envLocal_01);
    insertEnvironment(u, hello, envLocal_01);
    printEnvironment(stdout, envLocal_01, true);
    printEnvironment(stdout, envLocal_01, false);

    // Nest another environment inside envLocal_01
    Lexeme *i = newLexeme(ID_TYPE, "i");
    Lexeme *iVal = newLexeme(INTEGER_TYPE, "3");
    Lexeme *j = newLexeme(ID_TYPE, "j");
    Lexeme *jVal = newLexeme(INTEGER_TYPE, "9");
    Lexeme *envLocal_02 = extendEnvironment(NULL, NULL, envLocal_01);
    insertEnvironment(i, iVal, envLocal_02);
    insertEnvironment(j, jVal, envLocal_02);
    printEnvironment(stdout, envLocal_02, true);
    printEnvironment(stdout, envLocal_02, false);

    // Update a value in a global variable
    Lexeme *newVal = newLexeme(STRING_TYPE, "THIS IS THE NEW VALUE, YAY!");
    updateEnvironment(u, newVal, envLocal_02);

    // Lookup the new value
    fprintf(stdout, "The updated value of hello is: ");
    printLexeme(stdout, lookupEnvironment(u, envLocal_02));
    fprintf(stdout, "\n");

    // Extend global environment
    Lexeme *envError = extendEnvironment(NULL, NULL, global);

    // Try to access out-of-scope identifier
    lookupEnvironment(u, envError);

    return 0;
}
