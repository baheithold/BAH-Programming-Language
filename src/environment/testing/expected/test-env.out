Creating new environment...
Adding variable x with value 1
Adding variable y with value 2
Adding variable z with value 3
The environment is:
Scope Level: 0
z : INTEGER_TYPE 3
y : INTEGER_TYPE 2
x : INTEGER_TYPE 1

Extending the environment with w:4 and u:"hello"
The local environment is:
Scope Level: 0
u : STRING_TYPE "hello"
w : INTEGER_TYPE 4

The environment is:
Scope Level: 0
u : STRING_TYPE "hello"
w : INTEGER_TYPE 4
Scope Level: 1
z : INTEGER_TYPE 3
y : INTEGER_TYPE 2
x : INTEGER_TYPE 1

The local environment is:
Scope Level: 0
j : INTEGER_TYPE 9
i : INTEGER_TYPE 3

The environment is:
Scope Level: 0
j : INTEGER_TYPE 9
i : INTEGER_TYPE 3
Scope Level: 1
u : STRING_TYPE "hello"
w : INTEGER_TYPE 4
Scope Level: 2
z : INTEGER_TYPE 3
y : INTEGER_TYPE 2
x : INTEGER_TYPE 1

The updated value of hello is: STRING_TYPE "THIS IS THE NEW VALUE, YAY!"
